### ISSUE TITLE
AMQChannel.exnWrappingRpc should use a timeout

### ISSUE THREAD

**Issue by voronaam** [rabbitmq/rabbitmq-java-client#295]
It is possible to get RabbitMQ client in a "locked" state waiting for a response forever.

Steps to reproduce:
1. Have an application that publishes to a rabbitmq queue
2. Use `rabbitmqctl set_vm_memory_high_watermark 0` to block publishers on the server side
3. Start an application and make it attempt to publish something 

Expected result:
After some time, an timeout exception is thrown

Actual Result:
The produce is stuck forever.

Same callstacks captured from an application being stuck in this state:

```
"dwrr-thread-28" #56 prio=5 os_prio=0 tid=0x00007f4cb6710800 nid=0x579d in Object.wait() [0x00007f4c91481000]
   java.lang.Thread.State: WAITING (on object monitor)
        at java.lang.Object.wait(Native Method)
        - waiting on <0x00000007227801d0> (a com.rabbitmq.utility.BlockingValueOrException)
        at java.lang.Object.wait(Object.java:502)
        at com.rabbitmq.utility.BlockingCell.get(BlockingCell.java:49)
        - locked <0x00000007227801d0> (a com.rabbitmq.utility.BlockingValueOrException)
        at com.rabbitmq.utility.BlockingCell.uninterruptibleGet(BlockingCell.java:88)
        - locked <0x00000007227801d0> (a com.rabbitmq.utility.BlockingValueOrException)
        at com.rabbitmq.utility.BlockingValueOrException.uninterruptibleGetValue(BlockingValueOrException.java:32)
        at com.rabbitmq.client.impl.AMQChannel$BlockingRpcContinuation.getReply(AMQChannel.java:360)
        at com.rabbitmq.client.impl.AMQChannel.privateRpc(AMQChannel.java:225)
        at com.rabbitmq.client.impl.AMQChannel.exnWrappingRpc(AMQChannel.java:117)
        at com.rabbitmq.client.impl.ChannelN.queueBind(ChannelN.java:944)
        at com.rabbitmq.client.impl.ChannelN.queueBind(ChannelN.java:60)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.amqp.rabbit.connection.CachingConnectionFactory$CachedChannelInvocationHandler.invoke(CachingConnectionFactory.java:917)
        - locked <0x000000071ca91b50> (a java.lang.Object)
        at com.sun.proxy.$Proxy99.queueBind(Unknown Source)
        at org.springframework.amqp.rabbit.core.RabbitAdmin.declareBindings(RabbitAdmin.java:618)
        at org.springframework.amqp.rabbit.core.RabbitAdmin.access$300(RabbitAdmin.java:66)
        at org.springframework.amqp.rabbit.core.RabbitAdmin$8.doInRabbit(RabbitAdmin.java:296)
        at org.springframework.amqp.rabbit.core.RabbitTemplate.doExecute(RabbitTemplate.java:1436)
        at org.springframework.amqp.rabbit.core.RabbitTemplate.execute(RabbitTemplate.java:1392)
        at org.springframework.amqp.rabbit.core.RabbitTemplate.execute(RabbitTemplate.java:1368)
        at org.springframework.amqp.rabbit.core.RabbitAdmin.declareBinding(RabbitAdmin.java:293)
```

```
"dwrr-thread-3" #31 prio=5 os_prio=0 tid=0x00007f6c7a667800 nid=0x70ab in Object.wait() [0x00007f6c574f8000]
   java.lang.Thread.State: WAITING (on object monitor)
        at java.lang.Object.wait(Native Method)
        - waiting on <0x0000000727ce9c88> (a com.rabbitmq.utility.BlockingValueOrException)
        at java.lang.Object.wait(Object.java:502)
        at com.rabbitmq.utility.BlockingCell.get(BlockingCell.java:49)
        - locked <0x0000000727ce9c88> (a com.rabbitmq.utility.BlockingValueOrException)
        at com.rabbitmq.utility.BlockingCell.uninterruptibleGet(BlockingCell.java:88)
        - locked <0x0000000727ce9c88> (a com.rabbitmq.utility.BlockingValueOrException)
        at com.rabbitmq.utility.BlockingValueOrException.uninterruptibleGetValue(BlockingValueOrException.java:32)
        at com.rabbitmq.client.impl.AMQChannel$BlockingRpcContinuation.getReply(AMQChannel.java:360)
        at com.rabbitmq.client.impl.AMQChannel.privateRpc(AMQChannel.java:225)
        at com.rabbitmq.client.impl.AMQChannel.exnWrappingRpc(AMQChannel.java:117)
        at com.rabbitmq.client.impl.ChannelN.open(ChannelN.java:124)
        at com.rabbitmq.client.impl.ChannelManager.createChannel(ChannelManager.java:142)
        at com.rabbitmq.client.impl.AMQConnection.createChannel(AMQConnection.java:506)
        at org.springframework.amqp.rabbit.connection.SimpleConnection.createChannel(SimpleConnection.java:49)
        at org.springframework.amqp.rabbit.connection.CachingConnectionFactory$ChannelCachingConnectionProxy.createBareChannel(CachingConnectionFactory.java:1079)
        at org.springframework.amqp.rabbit.connection.CachingConnectionFactory$ChannelCachingConnectionProxy.access$500(CachingConnectionFactory.java:1068)
        at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.doCreateBareChannel(CachingConnectionFactory.java:521)
        at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.createBareChannel(CachingConnectionFactory.java:504)
        at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.getCachedChannelProxy(CachingConnectionFactory.java:474)
        at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.getChannel(CachingConnectionFactory.java:467)
        at org.springframework.amqp.rabbit.connection.CachingConnectionFactory.access$1500(CachingConnectionFactory.java:97)
        at org.springframework.amqp.rabbit.connection.CachingConnectionFactory$ChannelCachingConnectionProxy.createChannel(CachingConnectionFactory.java:1084)
        at org.springframework.amqp.rabbit.core.RabbitTemplate.doExecute(RabbitTemplate.java:1416)
        at org.springframework.amqp.rabbit.core.RabbitTemplate.execute(RabbitTemplate.java:1392)
        at org.springframework.amqp.rabbit.core.RabbitTemplate.send(RabbitTemplate.java:700)
        at org.springframework.amqp.rabbit.core.RabbitTemplate.convertAndSend(RabbitTemplate.java:768)
        at org.springframework.amqp.rabbit.core.RabbitTemplate.convertAndSend(RabbitTemplate.java:764)
```

Important notes:
A really outdated comment on https://github.com/rabbitmq/rabbitmq-java-client/blob/master/src/main/java/com/rabbitmq/client/impl/AMQChannel.java#L132 
The comment is 9 years old and the issue was fixed 6 years ago.

`AMQChannel` class has `rpc` and `privateRpc` versions with timeouts. Those should be used in `exnWrappingRpc`

Ideally, the timeout should be configurable.

**Comment by michaelklishin** [comment]
Thank you for your time.

Team RabbitMQ uses GitHub issues for specific actionable items engineers can work on. This assumes two things:

1. GitHub issues are not used for questions, investigations, root cause analysis, discussions of potential issues, etc (as defined by this team)
2. We have a certain amount of information to work with

We get at least a dozen of questions through various venues every single day, often quite light on details.
At that rate GitHub issues can very quickly turn into a something impossible to navigate and make sense of even for our team. Because of that questions, investigations, root cause analysis, discussions of potential features are all considered to be mailing list material by our team. Please post this to [rabbitmq-users](https://groups.google.com/forum/#!forum/rabbitmq-users).

Getting all the details necessary to reproduce an issue, make a conclusion or even form a hypothesis about what's happening can take a fair amount of time. Our team is multiple orders of magnitude smaller than the RabbitMQ community. Please help others help you by providing a way to reproduce the behavior you're
observing, or at least sharing as much relevant information as possible [on the list](https://groups.google.com/forum/#!forum/rabbitmq-users):

 * Server, client library and plugin (if applicable) versions used
 * Server logs
 * A code example or terminal transcript that can be used to reproduce
 * **Full** exception stack traces (not a single line message)
 * `rabbitmqctl status` (and, if possible, `rabbitmqctl environment` output)
 * Other relevant things about the environment and workload, e.g. [a traffic capture](https://www.rabbitmq.com/amqp-wireshark.html)

Feel free to edit out hostnames and other potentially sensitive information.

When/if we have enough details and evidence we'd be happy to file a new issue.

Thank you.

**Comment by michaelklishin** [comment]
You can set both RPC continuation timeout for channels and heartbeat timeout for connections really high. I would not recommend "infinite" values. Use [Publisher Confirms](http://www.rabbitmq.com/confirms.html) and make your publishers ready to re-publish on a timeout instead.

**Comment by michaelklishin** [comment]
Somehow this issue's title in my inbox is different and stated a somewhat different question. Anyhow, continuation timeouts were added in #219 and made safer in #290. Please upgrade to `4.2.0`.
